### ISSUE TITLE
Not detecting number plate though average loss is less than 0.06

### ISSUE THREAD

**Issue by NamburiSrinath** [AlexeyAB/darknet#2048]
Hi,
I am trying to detect number plate region given a car. The detailed problem has been described here(**as it is too lengthy)**. Please look into it and any help is very much appreciated.

[https://medium.com/@namburisrinath/hey-hi-80660945331b](url)

If any further clarifications/doubts are there, please do ask.

Thanks in advance
Srinath

**Comment by NamburiSrinath** [comment]
Thank you so much Alexey for this lucid explanation.

 So you are saying to train it with images which look like this
![1_xlabr2x5uo4jwbmmuh-qbw](https://user-images.githubusercontent.com/40389487/49940306-26763300-ff05-11e8-8d6e-87da1194a468.png)


rather than the number plate image directly.(which I have posted earlier)

Also, should I need to include any negative samples while training(like, hey, this is not a number plate!!!)?? Or YOLO takes care of that

Thanks in advance

**Comment by AlexeyAB** [comment]
>  So you are saying to train it with images which look like this
> ![1_xlabr2x5uo4jwbmmuh-qbw](https://user-images.githubusercontent.com/40389487/49940306-26763300-ff05-11e8-8d6e-87da1194a468.png)

Yes.

And with images like this:

![49936274-90caab80-fee4-11e8-8b34-d3dd1eaa1e30](https://user-images.githubusercontent.com/4096485/49941772-84e6e580-fef4-11e8-9a6c-4ca7b64520d6.jpeg)

----

> Also, should I need to include any negative samples while training(like, hey, this is not a number plate!!!)?? 

Yes. This is optional, but this is very desirable: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> desirable that your training dataset include images with non-labeled objects that you do not want to detect - negative samples without bounded box (empty .txt files) - use as many images of negative samples as there are images with objects

**Comment by NamburiSrinath** [comment]
Thanks again,

Small clarification Alexey.

What is the difference between first and second image that you have posted. Is it the clarity, position, rotation that's concerned?

Also, if I give negative samples the text file should be empty, that's fine.
Still the classes =1 and no need to add any new label names in obj.names file right. 

Negative samples means I can take any random images and place it in the training folder right. Any constraints on number or anything

Thanks in advance
Srinath

**Comment by AlexeyAB** [comment]
> What is the difference between first and second image that you have posted. Is it the clarity, position, rotation that's concerned?

As I see

for the 1st image `train_obj_width / train_image_width ~= 0.3`
for the 2nd image `train_obj_width / train_image_width ~= 0.15`

The more Training images similar to the Detection images - the better.

----


> Also, if I give negative samples the text file should be empty, that's fine.
> Still the classes =1 and no need to add any new label names in obj.names file right.
> 
> Negative samples means I can take any random images and place it in the training folder right. Any constraints on number or anything


Yes, just add any images. 

Then, for example, run https://github.com/AlexeyAB/Yolo_mark and go through all the images by hold SPACE button, so empty txt-files will be created automatically and these images will be added to the train.txt.

**Comment by AlexeyAB** [comment]
@NamburiSrinath 

> 1. Is it correct/okay if the negative images also have number plate attached to them?

It is very bad. Negative image must not have any objects. Because all object must be labeled, but negative images doesn't have labels: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> check that each object are mandatory labeled in your dataset - no one object in your data set should not be without label. In the most training issues - there are wrong labels in your dataset (got labels by using some conversion script, marked with a third-party tool, ...). Always check your dataset by using: https://github.com/AlexeyAB/Yolo_mark

----

> 2.Here are some of the screenshots that I attach when it was training. My dataset has only 1900 images but what does the number indicate there(83072 images, 66816 images). If it is data-augmentation please be specific as what effects are made on image and how many images it will produce etc.

Yes, it is data augmentation.
There will be generated infinity number of changed images.
Will be changed randomly: color, size and aspect ratio. (params in cfg file: `hue, exposure, saturation, jitter, random`)

----


> 3.How to improve the detection? The average loss error came less than 0.06xx as mentioned in your repo and it is detecting a few number plates correctly. But I want to improve its detection as
![img_3008_120_0](https://user-images.githubusercontent.com/40389487/50081201-e28c7200-0213-11e9-805a-aa4b24553d0a.jpg)

* **Use `yolov3.cfg` instead of `yolov2.cfg`.** It is much more accurate.

* Increase width=608 height=608 or width=832 height=832 in your cfg-file, then train and detect.

* Increase your training dataset. Include **many** images with such number plates to your Training dataset.

![img_3030_137_0](https://user-images.githubusercontent.com/40389487/50081202-e28c7200-0213-11e9-881c-15c991da6307.jpg)

![img_3224_151_0](https://user-images.githubusercontent.com/40389487/50081203-e28c7200-0213-11e9-82e1-8c781c705678.jpg)

![img_87_5_0](https://user-images.githubusercontent.com/40389487/50081204-e3250880-0213-11e9-8402-c9f47a96c98a.jpg)

**Comment by izesaon** [comment]
> resize

is resizing the same as cropping from the image to get the desired size? or resizing is simply resizing the image to the desired size while maintaining aspect ratio? i been trying to figure out why my model is not detecting properly- the loss is decreasing but so is the obj score...

**Comment by dis-is-pj** [comment]
How do I crop the prediction? I'm also working on number plate detection.

**Comment by NamburiSrinath** [comment]
> > resize
> 
> is resizing the same as cropping from the image to get the desired size? or resizing is simply resizing the image to the desired size while maintaining aspect ratio? i been trying to figure out why my model is not detecting properly- the loss is decreasing but so is the obj score...

No!!

Resizing and cropping the image are two different aspects. 

Assume you have an image of a car

1. Suppose, in the given example to detect number plates, if you crop the number plates)(from the car image) and feed it to the architecture, then your model will not recognize number plates when you give a test image(which has a car and a number plate)

2. If you resize the image(which has car and number plate), label it properly and feed it to network, then your model will detect a number plate when a test image is passed(which has a car and a number plate)

So, these two are different aspects. In general, while you are cropping, you are eliminating data from the image but in resizing you are not doing so.

**Note:** DNN Architectures do have constraints on the size of the image. So, always you can't expect them to detect a small football from a satellite image.(it depends though, but in my case YOLO is not **very good** for detecting small objects)

**Comment by NamburiSrinath** [comment]
> How do I crop the prediction? I'm also working on number plate detection.

Hello dis-is-pj

If you are working on Alexey AB, you can save the coordinates outputted by the network into a file. You can write a small script that takes the image as input(READ IMAGE) and crops it using these coordinates which you get from architecture. There are direct functions in OpenCV to do that. It's fairly simple (Check the format ---> image[a:b, c:d])
